shell/platform/linux/fl_basic_message_channel_test.cc

```diff
@@ -0,0 +1,146 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Included first as it collides with the X11 headers
+#include "gtest/gtest.h"
+
+#include "flutter/shell/platform/linux/fl_binary_messenger_private.h"
+#include "flutter/shell/platform/linux/fl_engine_private.h"
+#include "flutter/shell/platform/linux/public/flutter_linux/fl_basic_message_channel.h"
+#include "flutter/shell/platform/linux/public/flutter_linux/fl_standard_message_codec.h"
+#include "flutter/shell/platform/linux/testing/mock_renderer.h"
+
+// Creates a mock engine that responds to platform messages
+static FlEngine* make_mock_engine() {
+  g_autoptr(FlDartProject) project = fl_dart_project_new("data");
+  g_autoptr(FlMockRenderer) renderer = fl_mock_renderer_new();
+  g_autoptr(FlEngine) engine = fl_engine_new(project, FL_RENDERER(renderer));
+  g_autoptr(GError) error = nullptr;
+  EXPECT_TRUE(fl_engine_start(engine, &error));
+  EXPECT_EQ(error, nullptr);
+
+  return static_cast<FlEngine*>(g_object_ref(engine));
+}
+
+static void string_response_cb(GObject* object,
+                               GAsyncResult* result,
+                               gpointer user_data) {
+  g_autoptr(GError) error = nullptr;
+  g_autoptr(FlValue) message = fl_basic_message_channel_send_finish(
+      FL_BASIC_MESSAGE_CHANNEL(object), result, &error);
+  EXPECT_NE(message, nullptr);
+  EXPECT_EQ(error, nullptr);
+
+  EXPECT_EQ(fl_value_get_type(message), FL_VALUE_TYPE_STRING);
+  EXPECT_STREQ(fl_value_get_string(message), "Hello World!");
+
+  g_main_loop_quit(static_cast<GMainLoop*>(user_data));
+}
+
+// Checks can send a message and receive the response
+TEST(FlBasicMessageChannelTest, SendMessage) {
+  g_autoptr(GMainLoop) loop = g_main_loop_new(nullptr, 0);
+
+  g_autoptr(FlEngine) engine = make_mock_engine();
+  FlBinaryMessenger* messenger = fl_binary_messenger_new(engine);
+  g_autoptr(FlStandardMessageCodec) codec = fl_standard_message_codec_new();
+  g_autoptr(FlBasicMessageChannel) channel = fl_basic_message_channel_new(
+      messenger, "test/echo", FL_MESSAGE_CODEC(codec));
+  g_autoptr(FlValue) message = fl_value_new_string("Hello World!");
+  fl_basic_message_channel_send(channel, message, nullptr, string_response_cb,
+                                loop);
+  g_main_loop_run(loop);
+}
+
+static void failure_response_cb(GObject* object,
```

Comment:
Why are these interspersed with the tests? It seems like it would be easier to follow if they were grouped at the top.

These could also use some comments explaining what they do, test-wise.

Reply:
The file is structured:
```
<callbacks1>
<test case1>

<callbacks2>
<test case2>

<callbacks3>
<test case3>
```

It seems more confusing for all the callbacks to be grouped together like:
```
<callbacks1>
<callbacks2>
<callbacks3>
<test case1>
<test case2>
<test case3>
```

Reply:
I see, I thought things were being re-used later. Makes sense.